atomic_reactor/plugins/pre_fetch_maven_artifacts.py

```diff
@@ -0,0 +1,216 @@
+        :param koji_krb_keytab: str, Kerberos keytab
+        :param allowed_domains: list<str>: list of domains that are
+               allowed to be used when fetching artifacts by URL (case insensitive)
+        """
+        super(FetchMavenArtifactsPlugin, self).__init__(tasker, workflow)
+        koji_auth = {
+            'proxyuser': koji_proxyuser,
+            'ssl_certs_dir': koji_ssl_certs_dir,
+            'krb_principal': koji_krb_principal,
+            'krb_keytab': koji_krb_keytab,
+        }
+        # Remove empty values from auth dict to avoid login when not needed
+        koji_auth = {k: v for k, v in koji_auth.items() if v}
+        self.koji_info = {
+            'hub': koji_hub,
+            'root': koji_root,
+            'auth': koji_auth or None
+        }
+        self.path_info = PathInfo(topdir=self.koji_info['root'])
+        self.allowed_domains = set(domain.lower() for domain in allowed_domains or [])
+        self.workdir = self.workflow.source.get_dockerfile_path()[1]
+
+    def read_nvr_requests(self):
+        file_path = os.path.join(self.workdir, self.NVR_REQUESTS_FILENAME)
+        if not os.path.exists(file_path):
+            self.log.debug('%s not found', self.NVR_REQUESTS_FILENAME)
+            return []
+
+        nvr_requests = util.read_yaml(file_path, default=[],
+                                      schema='schemas/fetch-artifacts-nvr.json')
+        return [NvrRequest(**nvr_request) for nvr_request in nvr_requests]
+
+    def read_url_requests(self):
+        file_path = os.path.join(self.workdir, self.URL_REQUESTS_FILENAME)
+        if not os.path.exists(file_path):
+            self.log.debug('%s not found', self.URL_REQUESTS_FILENAME)
+            return []
+
+        return util.read_yaml(file_path, default=[],
+                              schema='schemas/fetch-artifacts-url.json')
+
+    def process_by_nvr(self, nvr_requests):
+        download_queue = []
+
+        for nvr_request in nvr_requests:
+            build_info = self.session.getBuild(nvr_request.nvr)
+            if not build_info:
+                raise ValueError('Build {} not found.'.format(nvr_request.nvr))
+
+            maven_build_path = self.path_info.mavenbuild(build_info)
+            build_archives = self.session.listArchives(buildID=build_info['id'],
+                                                       type='maven')
+            build_archives = nvr_request.match_all(build_archives)
+
+            for build_archive in build_archives:
+                maven_file_path = self.path_info.mavenfile(build_archive)
+                # NOTE: Don't use urljoin here because maven_build_path does
+                # not contain a traling slash, which causes the last dir to
+                # be dropped.
+                url = maven_build_path + '/' + maven_file_path
+                download_queue.append(DownloadRequest(url, maven_file_path,
+                                                      build_archive['checksum']))
```

Comment:
Would there be any benefit in making this a generator function? (Although, not sure how would ever collect all errors in that case.)